client/session.go

```diff
@@ -2816,17 +2823,22 @@ func (b *baseBlocksResult) newDatabaseBlock(block *rpc.Block) (block.DatabaseBlo
 	case segments.Unmerged != nil:
 		// Must merge to provide a single block
 		segmentReaderPool := b.blockOpts.SegmentReaderPool()
-		readers := make([]io.Reader, len(segments.Unmerged))
-		for i := range segments.Unmerged {
+		readers := make([]xio.Reader, len(segments.Unmerged))
+		end := time.Time{}
+		for i, segment := range segments.Unmerged {
 			segmentReader := segmentReaderPool.Get()
-			segmentReader.Reset(b.segmentForBlock(segments.Unmerged[i]))
-			readers[i] = segmentReader
+			endTime := timeConvert(segment.EndTime)
+			blockReader := xio.NewBlockReader(segmentReader, timeConvert(segment.StartTime), endTime)
+			blockReader.Reset(b.segmentForBlock(segment))
+			readers[i] = blockReader
+			if end.Before(endTime) {
+				end = endTime
+			}
 		}
-
-		encoder, err := b.mergeReaders(start, readers)
+		encoder, err := b.mergeReaders(start, end, readers)
 		for _, reader := range readers {
 			// Close each reader
-			segmentReader := reader.(xio.SegmentReader)
+			segmentReader := reader.(xio.BlockReader)
```

Comment:
Perhaps rename this to `blockReader := ...`

Reply:
still applicable?

Reply:
Must have accidentally undone the fix